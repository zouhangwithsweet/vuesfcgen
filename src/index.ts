import { baseParse, TemplateChildNode } from '@vue/compiler-core'
import { transform } from '@babel/core'

const NORMAL_ATTR = 6

// babel transform script code
function transformScript(ast: TemplateChildNode, options?: any) {
  if ('tag' in ast) {
    if (ast.tag === 'script') {
      const sourceCode = ast.loc.source
      let code = ``
      code += transform(sourceCode, options)
      return code
    }
  }
  return ``
}

// todo transform style ast
function transformStyle(ast: TemplateChildNode, options?: any) {
  if ('tag' in ast) {
    if (ast.tag === 'style') {
      return ast.loc.source
    }
  }
  return ``
}

// gen template
function createCodegenContext(ast: TemplateChildNode) {
  const context = {
    code: ``,
    push(code: string) {
      context.code += code
    },
    source: ast.loc.source,
    newline(n = 2) {
      context.push('\n' + ` `.repeat(n))
    }
  }
  return context
}

function genTemplate(ast: TemplateChildNode) {
  const context = createCodegenContext(ast)
  const { push, newline } = context

  push('<template>')
  newline()

  genHTML(ast, context)

  push('</template>')
  newline()
  return context
}

function genHTML(ast: TemplateChildNode, context) {
  const { push, newline } = context
  if ('children' in ast) {
    ast.children.forEach(child => {
      const { tag, isSelfClosing, props } = child
      if (tag) {
        if (isSelfClosing) {
          push(`<${tag}`)
          injectProps(props, context)
          push(`/>`)
          newline()
        } else {
          push(`<${tag}`)
          injectProps(props, context)
          push(`>`)
          if (child.children && child.children.length) {
            genHTML(child, context)
          }
          push(`</${tag}>`)
          newline()
        }
      } else {
        push(child.loc.source)
      }
    })
  }
}

function injectProps(props, context) {
  const { push } = context
  props.forEach(prop => {
    push(` `)
    const { type, name, value, loc } = prop
    if (type === NORMAL_ATTR) {
      value
        ? push(`${name}=${value.loc.source}`)
        : push(`${name}`)
    } else {
      push(loc.source)
    }
  })
}
