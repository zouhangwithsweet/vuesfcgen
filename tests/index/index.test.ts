import fs from 'fs'
import path from 'path'
import {
  baseParse, RootNode,
} from '@vue/compiler-core'
import { transformScript, transformStyle } from '../../src/index'
const filePath = path.resolve(__dirname, './Index.vue');

const content = fs.readFileSync(filePath).toString();
const rootNode = baseParse(content);

test('transformScript', () => {
  const scriptNode = rootNode.children.find(_ => 'tag' in _ && _.tag === 'script')
  if (scriptNode) {
      const result = transformScript(scriptNode, { filename: 'file.ts' });
      expect(result).toMatchSnapshot();
  }
  expect(true).toBe(true);
});

// TODO
test.skip('transformStyle', () => {
  const scriptNode = rootNode.children.find(_ => 'tag' in _ && _.tag === 'style')
  if (scriptNode) {
      const result = transformStyle(scriptNode);
      console.log(result);
      expect(true).toBe(true);
  }
  expect(true).toBe(true);
});
