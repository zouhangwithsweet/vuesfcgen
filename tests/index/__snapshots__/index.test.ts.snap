// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformScript 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _default = {
  data() {
    return {
      greeting: 'Hello'
    };
  }

};
exports.default = _default;"
`;
